### PR TITLE
fix(webgl): Apply default viewport sizes

### PR DESCRIPTION
**Pull request by ibgreen**
#### Background
- Viewport was not set by default, causing examples to not fill screen

#### Change List
- Apply reasonable defaults to viewports when not supplied.
- Document behavior

Before
image

After
image

### CODE REVIEW

File: docs/api-reference/core/resources/render-pass.md [hunk 1]
```diff
@@ -45,6 +45,14 @@ Depth and stencil buffers are also cleared to default values:
   device.submit();
 ```
 
+## Viewport size
+
+`RenderPassProps.parameters.viewport` controls how the rendered graphics is mapped to window pixels / texels (more precisely, the affine transformation of x and y from normalized device coordinates to window coordinates).
+
+If no value for the `viewport` parameter is provided, the following defaults will be applied.
+- If no `framebuffer` is specified, the size of the canvas drawing buffer will be used (`[gl.canvas.drawingBufferWidth, gl.canvas.drawingBufferHeight]`)
```
Comment (chrisgervang) [inline docs/api-reference/core/resources/render-pass.md:53]: Is there a WebGPU equivalent?

File: modules/webgl/src/adapter/converters/texture-formats.ts [hunk 2]
```diff
@@ -509,7 +509,7 @@ export function getTextureFormatSupport(
 
   // Support Check that we have a GL constant
   let supported = info.gl === undefined;
-  supported = supported && checkTextureFeature(gl, info.f as DeviceFeature, extensions);
+  supported = supported && checkTextureFeature(gl, info.f, extensions);
```
Comment (chrisgervang) [inline modules/webgl/src/adapter/converters/texture-formats.ts:512]: Hard to tell in the diff, is this related?

**Review by chrisgervang — APPROVED** [review]

**Comment by chrisgervang** [comment]
LGTM assuming these values are post-application of device pixel ratio.

File: modules/webgl/src/adapter/resources/webgl-render-pass.ts [hunk 3]
```diff
@@ -28,9 +28,22 @@ export class WEBGLRenderPass extends RenderPass {
...
+    // If no viewport is provided, apply reasonably defaults
+    let viewport;
+    if (!props?.parameters?.viewport) {
+      if (props?.framebuffer) {
+        // Set the viewport to the size of the framebuffer
+        const {width, height} = props.framebuffer;
+        viewport = [0, 0, width, height];
+      } else {
+        // Instead of using our own book-keeping, we can just read the values from the WebGL context
+        viewport = [0, 0, device.gl.drawingBufferWidth, device.gl.drawingBufferHeight];
```
Comment (donmccurdy) [inline modules/webgl/src/adapter/resources/webgl-render-pass.ts:40]: Optional:

```suggestion
        viewport = [0, 0, ...device.getSize()];
```

**Review by donmccurdy — APPROVED** [review]

File: modules/webgl/src/adapter/converters/texture-formats.ts (hunk 2 above)
Comment (ibgreen) [reply modules/webgl/src/adapter/converters/texture-formats.ts:512]: Huh, no, must be some some lint/prettier autofix. But if the casts are not needed, then better drop them...

File: docs/api-reference/core/resources/render-pass.md (hunk 1 above)
Comment (ibgreen) [reply docs/api-reference/core/resources/render-pass.md:53]: > Is there a WebGPU equivalent?

Very good point, the docs should not be WebGL specific. Will drop.

File: modules/webgl/src/adapter/resources/webgl-render-pass.ts (hunk 3 above)
Comment (ibgreen) [reply modules/webgl/src/adapter/resources/webgl-render-pass.ts]: > viewport = [0, 0, ...device.getSize()];

I used `device.getCanvasContext().getDrawingBufferSize()`.  I am not sure about why have a `device.getSize()` - it seems it is only available on WebGLDevice and NullDevice. I will put up a separate PR to propose removing it.
